hadoop-hdds/server-scm/src/main/java/org/apache/hadoop/hdds/scm/container/replication/LegacyReplicationManager.java

```diff
@@ -102,6 +105,111 @@ public class LegacyReplicationManager {
   public static final Logger LOG =
       LoggerFactory.getLogger(LegacyReplicationManager.class);
 
+  static class InflightMap {
+    private final Map<ContainerID, List<InflightAction>> map
+        = new ConcurrentHashMap<>();
+    private final InflightType type;
+    private final int sizeLimit;
+    private final AtomicInteger inflightCount = new AtomicInteger();
+
+    InflightMap(InflightType type, int sizeLimit) {
+      this.type = type;
+      this.sizeLimit = sizeLimit > 0 ? sizeLimit : Integer.MAX_VALUE;
+    }
+
+    boolean isReplication() {
+      return type == InflightType.REPLICATION;
+    }
+
+    private List<InflightAction> get(ContainerID id) {
+      return map.get(id);
+    }
+
+    boolean containsKey(ContainerID id) {
+      return map.containsKey(id);
+    }
+
+    int inflightActionCount(ContainerID id) {
+      return Optional.ofNullable(map.get(id)).map(List::size).orElse(0);
+    }
+
+    int containerCount() {
+      return map.size();
+    }
+
+    boolean isFull() {
+      return inflightCount.get() >= sizeLimit;
+    }
+
+    void clear() {
+      map.clear();
+    }
+
+    void iterate(ContainerID id, Function<InflightAction, Boolean> processor) {
+      for (; ;) {
+        final List<InflightAction> actions = get(id);
+        if (actions == null) {
+          return;
+        }
+        synchronized (actions) {
+          if (get(id) != actions) {
+            continue; //actions is changed, retry
+          }
+          for (Iterator<InflightAction> i = actions.iterator(); i.hasNext();) {
+            final Boolean remove = processor.apply(i.next());
+            if (remove == Boolean.TRUE) {
```

Comment:
> The windbags failure does not seem related to this.

`M B RC: Suspicious comparison of Boolean references in org.apache.hadoop.hdds.scm.container.replication.LegacyReplicationManager$InflightMap.iterate(ContainerID, Function)  At LegacyReplicationManager.java:[line 160]`

I think it is related, assuming windbags == findbugs. :)

Maybe I'm missing something, but can we use `Predicate<InflightAction>` instead of `Function<InflightAction, Boolean>`?

Reply:
Also, Predicate sounds great!

Reply:
The workflow has a "summary of failures" section which greps for the real findbugs problems, to avoid the need for checking the complete output manually:

https://github.com/apache/ozone/runs/6874432894#step:6:8

Reply:
Good to know.  Thanks!